### PR TITLE
Add foundation for importing ONNX files

### PR DESCRIPTION
**Pull request by antimora**
Checking the initial work for importing ONNX files. Now there is a limited ability to generate burn model source code from an onnx file. Currently only a few ops supported but sufficient for a working example.

### CODE REVIEW

File: burn-import/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        .out_dir("model/")            // Directory to output the generated Rust source file (under target/)
+        .run_from_script();
+}
+```
+
+Then, add the following to mod.rs under `src/model`:
+
+```rust
+pub mod mnist {
+    include!(concat!(env!("OUT_DIR"), "/model/mnist.rs"));
```
Comment (nathanielsimard) [inline burn-import/README.md:38]: We could provide an attributes that does everything during compile time.

```rust
#[burn(import = "src/model/mnist.onnx")]
struct ModelName<B>; // Going to be used as model name.
```

File: burn-import/src/onnx/codegen.rs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+        }
+    }
+
+    /// Generates source code for the stateless node calls, i.e. add, mul, etc.
+    fn node_call_stateless(node: &Node, imports: &mut HashSet<String>) -> TokenStream {
+        if node.is_stateful {
+            panic!("Node must be stateless");
+        }
+
+        // let name = Ident::new(&node.name, Span::call_site());
```
Comment (nathanielsimard) [inline burn-import/src/onnx/codegen.rs:320]: deep code

File: burn-import/src/onnx/codegen.rs [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+        let mut outputs = vec![];
+
+        for output in node.outputs.iter() {
+            let name = Ident::new(&output.name, Span::call_site());
+            outputs.push(quote! {
+                #name
+            });
+        }
+
+        // let input = inputs.pop().unwrap();
```
Comment (nathanielsimard) [inline burn-import/src/onnx/codegen.rs:340]: deep code

File: burn-import/src/onnx/protos/onnx.proto [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+
+// SPDX-License-Identifier: Apache-2.0
+
+
+syntax = "proto3";
+
+package onnx;
+
+// Overview
+//
```
Comment (nathanielsimard) [inline burn-import/src/onnx/protos/onnx.proto:17]: Why do we need the proto?

File: burn-import/README.md (hunk 1 above)
Comment (antimora) [reply burn-import/README.md:38]: That's a great idea. I can take this up in the next PRs. It should not be difficult to add this since the code is generated with TokenStream. I actually intended to use in macro, I was just trying to figure out a good syntax for it, which you provided just now.

File: burn-import/src/onnx/codegen.rs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+                #(#call_nodes)*
+                #results
+            }
+        }
+    }
+
+    /// Generates source code for the stateful node calls, i.e. conv, dropout, etc.
+    fn node_call_stateful(node: &Node) -> TokenStream {
+        if !node.is_stateful {
+            panic!("Node must be stateful");
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/codegen.rs:258]: Stateful as an opposite to functional like 'add', `mult`, `flatten` ops. The state I am referring to is weights, etc. I needed some differentiation so these nodes can declared. The states will be loaded from ONNX file.

File: burn-import/src/onnx/codegen.rs (hunk 2 above)
Comment (antimora) [reply burn-import/src/onnx/codegen.rs:320]: Yeah, I realize it but this is just an initial version. I imagine this can be refactored and made more modular.

File: burn-import/src/onnx/codegen.rs (hunk 3 above)
Comment (antimora) [reply burn-import/src/onnx/codegen.rs:340]: oh, you mean dead code. OK. I'll remove it.

File: burn-import/src/onnx/codegen.rs [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+        let mut fields = vec![];
+
+        for node in self.graph.nodes.iter().filter(|x| x.is_stateful) {
+            let node_type = match node.node_type {
+                NodeType::Conv1d => syn::parse_str::<Type>("nn::conv::Conv1d<B>").unwrap(),
+                NodeType::Conv2d => syn::parse_str::<Type>("nn::conv::Conv2d<B>").unwrap(),
+                NodeType::Linear => syn::parse_str::<Type>("nn::Linear<B>").unwrap(),
+                _ => {
+                    todo!("Node type not implemented: {:?}", node.node_type)
+                }
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/codegen.rs:480]: Conv1d, Conv2d and Linear ops are not present in ONNX spec. So we convert/collapse into these nodes .

File: burn-import/src/onnx/codegen.rs [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+
+    /// Generates source code for the initialization method
+    fn initialization_fields(&self) -> Vec<TokenStream> {
+        let mut fields = vec![];
+
+        for node in self.graph.nodes.iter().filter(|x| x.is_stateful) {
+            let init_code = match node.node_type {
+                NodeType::Conv2d => conv2d_init(node),
+                NodeType::Linear => linear_init(node),
+                _ => {
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/codegen.rs:499]: Yeah, probably. This is just an initial version. I wanted to see what pattern emerge before generalizing this. Probably some trait on Enum could be defined and implemented in the same module.

File: burn-import/src/onnx/convert.rs [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+pub fn parse_onnx(onnx_path: &Path) -> Graph {
+    let mut file = File::open(onnx_path).expect("Unable to open file");
+    let onnx_model: ModelProto =
+        Message::parse_from_reader(&mut file).expect("Unable to parse ONNX file");
+
+    // Convert the nodes
+    let mut nodes: Vec<Node> = vec![];
+    for onnx_node in onnx_model.graph.node.iter() {
+        nodes.push(convert_node_proto(onnx_node));
+    }
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/convert.rs:46]: It's more useful doing in Vec because the nodes need to be sorted topologically (keyed by node inputs/outputs). There is currently no use case to loop up by node's name, only by inputs and outputs. 

Initially I was using map but it added up being more difficult to use.

File: burn-import/src/onnx/convert.rs [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+];
+
+#[derive(Debug)]
+/// Error type for parsing ONNX model
+pub enum ParseError {
+    VariantNotFound,
+}
+
+/// Open an onnx file and convert it to a Graph (intermediate representation)
+pub fn parse_onnx(onnx_path: &Path) -> Graph {
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/convert.rs:37]: OK. I'll extra some parts into function.

File: burn-import/src/onnx/ir.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+/// The list of supported node types (ONNX operators and some extra ones)
+#[derive(Debug, Hash, Eq, PartialEq, EnumString, Clone, Display)]
+pub enum NodeType {
+    // Additional Burn operators
+    Linear,
+    Gelu,
+    Conv1d,
+    Conv2d,
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/ir.rs:133]: There is a little use for it in the code. It's nice that we know some node types do not exist in ONNX spec but for the code this separation is not need. Even we decide on using this IR for exporting, the conversion will be handled by code. Lastly, it'd be difficult to check for ONNX vs BURN op type.

File: burn-import/src/onnx/shape_inference.rs [hunk 11]
```diff
@@ -0,0 +1,187 @@
+use std::collections::HashMap;
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/shape_inference.rs:1]: The ONNX spec makes the input shapes implicit. You only figure out by passing a concrete input from the top. For us, in order to declare Conv or Linear, we need to know in/out channels. This information is absent from the spec. So you have to figure out by look at the input shape or get from weights. To be honest, it's bad that ONNX omitted from storing in the file.

File: burn-import/src/onnx/ir.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply burn-import/src/onnx/ir.rs]: Well I think it's important to avoid as much as possible using `_ => ...` in match statements. So if you only need Burn operation in one function, it's probably best to have multiple enum? Does it make sense?

File: burn-import/src/onnx/shape_inference.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply burn-import/src/onnx/shape_inference.rs:1]: That's terrible 😅

File: burn-import/src/onnx/ir.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/onnx/ir.rs]: I updated the enum and mixed all in one. I explained in the doc comment that these are mixed. To iterate my point there is no distinction for the code generation if it's burn or onnx specific node. In fact, I'll make it clear in the readme, the code generation only should care about IR data model. 

As far as the current implementation using matches, I agree it's not the best style with all catch but this is a simple first version. We will refactor when we can identify a better pattern.

**Comment by blefaudeux** [comment]
hey there, I'm dipping my toes in burn and was interested in this path (onnx import) for a start, I just stumbled onto this PR (disclaimer: very noob in Rust but some notions). As a proof-of-concept I played with importing a ResNet50, there are some issues of course but I was interested in your plans here ? Is there some reference somewhere, something which would convey the amount of work needed and current progress ? 

For reference, BatchNorm import was not supported (ok in a draft) + Conv expects a batch dimension (also fixed in a draft), now onto a Conv2D padding pattern (`not yet implemented: Conv2d: padding([3, 3, 3, 3]) is not fully supported`) but I'm wondering whether this is all known and duplicate work already :) 

Thanks for the great work in any case

**Comment by antimora** [comment]
Hi @blefaudeux, it's great to hear that you're interested in using this project! I'm currently in the process of adding support for loading weights from ONNX files. Once that's completed and merged into the main branch, I'll focus on tests and documentation to encourage more contributions. Over time, I plan to add more Burn operations, and it would be fantastic if more people could contribute.

There are numerous cases in ONNX, so my strategy has been to start with well-known and popular use cases (e.g., MNIST ONNX exported from PyTorch) and build up from there. That's why you may see some incomplete cases, such as those with padding. If we can get your model working, it would be a significant milestone for Burn. Any contribution is greatly appreciated!

I'm not sure if your ResNet50 model requires MeanPool2d (currently unavailable in Burn; see https://github.com/burn-rs/burn/issues/135), but I suspect it's relatively easy to add and similar to MaxPool2d. It probably doesn't require a backward pass.

@nathanielsimard has been working on adding the remaining features for Conv operations (strides and dilation are complete, and grouping is up next).

We have a live thread about ONNX for Burn on Discord: https://discord.com/channels/1038839012602941528/1088240873420095608 You're welcome to chime in.

**Comment by antimora** [comment]
@blefaudeux The state loading has landed on main. Please try out and give us some feedback. We have a couple outstanding tasks to further improve the import functionality. In the latest version, you'll be able to use BatchNorm op. Also in Burn we have added AveragePool2d.

**Comment by blefaudeux** [comment]
ohhh that's great to hear ! in the train atm but I'll give it a shot, really curious to see where this goes. I didn't submit my changes in the end (asked on discord), it was still in the middle of the fording and I figured you would do it faster. 
cc @MatthieuTPHR @david-PHR

**Comment by antimora** [comment]
@blefaudeux forgot to point to a working example: https://github.com/burn-rs/burn/tree/main/examples/onnx-inference

cc @MatthieuTPHR @david-PHR
